File: ios/AuraKit/Tests/AuraKitTests/AuraKitTests.swift
// AuraKitTests.swift — тесты общего слоя без сети: JSON, кадры протокола,
// deep links. Гоняются `swift test` в ios/AuraKit (или XCTest в Xcode).
import XCTest

@testable import AuraKit

final class JSONValueTests: XCTestCase {

    func testRoundTrip() throws {
        let json = """
        {"id": 12, "ok": true, "text": "привет", "tags": ["a", "b"], "meta": null}
        """
        let value = try JSONDecoder().decode(JSONValue.self, from Data(json.utf8))
        XCTAssertEqual(value.int("id"), 12)
        XCTAssertEqual(value.bool("ok"), true)
        XCTAssertEqual(value.string("text"), "привет")
        XCTAssertEqual(value.array("tags").count, 2)
        XCTAssertEqual(value["meta"], .null)

        let reEncoded = try JSONEncoder().encode(value)
        let decodedAgain = try JSONDecoder().decode(JSONValue.self, from: reEncoded)
        XCTAssertEqual(decodedAgain, value)
    }

    func testDefaults() {
        let value = JSONValue.object([:])
        XCTAssertEqual(value.string("missing", default: "x"), "x")
        XCTAssertEqual(value.int("missing", default: 7), 7)
        XCTAssertEqual(value.array("missing"), [])
    }
}

final class ProtocolFrameTests: XCTestCase {

    func testRequestFrameShape() throws {
        let frame = AuraRequestFrame(id: "1", type: "auth.login",
                                     payload: .object(["email": .string("a@b.c")]))
        let data = try JSONEncoder().encode(frame)
        let object = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        XCTAssertEqual(object?["id"] as? String, "1")
        XCTAssertEqual(object?["type"] as? String, "auth.login")
        let payload = object?["payload"] as? [String: Any]
        XCTAssertEqual(payload?["email"] as? String, "a@b.c")
    }

    func testErrorFrameDecoding() throws {
        let json = """
        {"id":"1","type":"error","code":"bad_request","message":"нужен email"}
        """
        let frame = try JSONDecoder().decode(AuraResponseFrame.self, from: Data(json.utf8))
        XCTAssertEqual(frame.type, "error")
        XCTAssertEqual(frame.code, "bad_request")
        XCTAssertEqual(frame.message, "нужен email")
    }

    func testEventFrameDecoding() throws {
        let json = """
        {"type":"event","event":"chat.message","payload":{"text":"ок"}}
        """
        let frame = try JSONDecoder().decode(AuraResponseFrame.self, from: Data(json.utf8))
        XCTAssertEqual(frame.type, "event")
        XCTAssertEqual(frame.event, AuraEventName.chatMessage)
        XCTAssertEqual(frame.payload?.string("text"), "ок")
    }
}

final class DeepLinkTests: XCTestCase {

    func testChats() {
        XCTAssertEqual(DeepLink(url: URL(string: "aura://chats/42")!), .chats(id: 42))
        XCTAssertEqual(DeepLink(url: URL(string: "aura://chats")!), .chats(id: nil))
    }

    func testSimpleRoutes() {
        XCTAssertEqual(DeepLink(url: URL(string: "aura://voice")!), .voice)
        XCTAssertEqual(DeepLink(url: URL(string: "aura://tasks")!), .tasks)
        XCTAssertEqual(DeepLink(url: URL(string: "aura://settings")!), .settings)
    }

    func testAsk() {
        let link = DeepLink(url: URL(string: "aura://ask?text=%D0%9F%D1%80%D0%B8%D0%B2%D0%B5%D1%82")!)
        XCTAssertEqual(link, .ask(text: "Привет"))
    }

    func testOAuthCallback() {
        let link = DeepLink(url: URL(string: "aura://oauth?provider=google_gmail&code=abc&state=st1")!)
        XCTAssertEqual(link, .oauth(provider: "google_gmail", code: "abc", state: "st1"))
    }

    func testRejectsForeignSchemeAndUnknownHost() {
        XCTAssertNil(DeepLink(url: URL(string: "https://aura.ai/chats/1")!))
        XCTAssertNil(DeepLink(url: URL(string: "aura://unknown")!))
        XCTAssertNil(DeepLink(url: URL(string: "aura://oauth?provider=x")!))  // нет code/state
    }
}

File: android/aurakit/src/test/kotlin/ai/aura/kit/AuraKitTests.kt
package ai.aura.kit

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

// AuraKitTests — тесты общего слоя без сети: кадры протокола, deep links,
// доступы к JsonObject. Запуск: ./gradlew :aurakit:test (JVM, без эмулятора).
// Зеркало ios/AuraKit/Tests/AuraKitTests/AuraKitTests.swift.

class ProtocolFrameTests {

    private val json = Json { ignoreUnknownKeys = true }

    @Test
    fun `request frame shape`() {
        val frame = AuraRequestFrame(
            id = "1",
            type = "auth.login",
            payload = buildJsonObject { put("email", "a@b.c") },
        )
        val encoded = json.encodeToString(AuraRequestFrame.serializer(), frame)
        val decoded = json.parseToJsonElement(encoded)
        val obj = decoded as kotlinx.serialization.json.JsonObject
        assertEquals("1", (obj["id"] as kotlinx.serialization.json.JsonPrimitive).content)
        assertEquals("auth.login", (obj["type"] as kotlinx.serialization.json.JsonPrimitive).content)
        val payload = obj["payload"] as kotlinx.serialization.json.JsonObject
        assertEquals("a@b.c", payload.str("email"))
    }

    @Test
    fun `error frame decoding`() {
        val frame = json.decodeFromString(
            AuraResponseFrame.serializer(),
            """{"id":"1","type":"error","code":"bad_request","message":"нужен email"}""",
        )
        assertEquals("error", frame.type)
        assertEquals("bad_request", frame.code)
        assertEquals("нужен email", frame.message)
    }

    @Test
    fun `event frame decoding`() {
        val frame = json.decodeFromString(
            AuraResponseFrame.serializer(),
            """{"type":"event","event":"chat.message","payload":{"body":"ок"}}""",
        )
        assertEquals("event", frame.type)
        assertEquals("chat.message", frame.event)
        assertEquals("ок", frame.payload.str("body"))
        assertNull(frame.id)
    }

    @Test
    fun `event names match protocol`() {
        // Имена событий — часть протокола (docs/PROTOCOL.md); сверяются здесь,
        // чтобы переименование на клиенте не разъехалось с сервером.
        assertEquals("chat.message", AuraEventName.CHAT_MESSAGE)
        assertEquals("task.due", AuraEventName.TASK_DUE)
        assertEquals("notification.new", AuraEventName.NOTIFICATION_NEW)
    }

    @Test
    fun `json accessors are null safe`() {
        val value: kotlinx.serialization.json.JsonObject? = null
        assertEquals("", value.str("missing"))
        assertEquals(0L, value.long("missing"))
        assertEquals(false, value.bool("missing"))
        assertTrue(value.arr("missing").isEmpty())

        val obj = buildJsonObject { put("n", 12); put("s", "привет"); put("nil", JsonNull) }
        assertEquals(12L, obj.long("n"))
        assertEquals("привет", obj.str("s"))
        assertEquals("", obj.str("nil"))
    }
}

class DeepLinkTests {

    @Test
    fun `chats route with and without id`() {
        assertEquals(DeepLink.Chats(42), DeepLink.parse("aura://chats/42"))
        assertEquals(DeepLink.Chats(null), DeepLink.parse("aura://chats"))
        assertEquals(DeepLink.Chats(null), DeepLink.parse("aura://chats/не-число"))
    }

    @Test
    fun `simple routes`() {
        assertEquals(DeepLink.Voice, DeepLink.parse("aura://voice"))
        assertEquals(DeepLink.Tasks, DeepLink.parse("aura://tasks"))
        assertEquals(DeepLink.Notifications, DeepLink.parse("aura://notifications"))
        assertEquals(DeepLink.Settings, DeepLink.parse("aura://settings"))
    }

    @Test
    fun `ask route decodes query`() {
        assertEquals(
            DeepLink.Ask("привет, Аура"),
            DeepLink.parse("aura://ask?text=%D0%BF%D1%80%D0%B8%D0%B2%D0%B5%D1%82%2C%20%D0%90%D1%83%D1%80%D0%B0"),
        )
        assertEquals(DeepLink.Ask(""), DeepLink.parse("aura://ask"))
    }

    @Test
    fun `oauth callback requires all three params`() {
        assertEquals(
            DeepLink.OAuth("google_calendar", "code-1", "state-1"),
            DeepLink.parse("aura://oauth?provider=google_calendar&code=code-1&state=state-1"),
        )
        assertNull(DeepLink.parse("aura://oauth?provider=google_calendar&code=code-1"))
    }

    @Test
    fun `rejects foreign scheme and unknown host`() {
        assertNull(DeepLink.parse("https://aura.app/chats/1"))
        assertNull(DeepLink.parse("aura://unknown"))
        assertNull(DeepLink.parse("не url"))
    }
}
